_posts/2020-07-06-IP_address_manager.md

```diff
@@ -0,0 +1,179 @@
+object, for the pivoting to proceed properly.
+
+In a nutshell, the manager provides an IP Address allocation service, based
+on Kubernetes API and fulfilling the needs of Metal3, specifically the
+requirements of CAPI.
+
+## How does it work ?
+
+The manager follows the same logic as the volume allocation in Kubernetes,
+with a claim and an object created for that claim. There are three type of
+objects defined, the IPPool, the IPClaim and the IPAddress objects.
+
+The IPPool objects contain the definition of the IP subnets from which the
+Addresses are allocated. It is IP version agnostic. The subnets can either
+be defined as such or given as a start and end IP addresses with a prefix.
+It also supports pre-allocating IP addresses.
+
+The following is an example IPPool definition :
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: pool1
+spec:
+  clusterName: cluster1
+  pools:
+    - start: 192.168.0.10
+      end: 192.168.0.30
+      prefix: 25
+      gateway: 192.168.0.1
+    - subnet: 192.168.1.1/26
+    - subnet: 192.168.1.128/25
+  prefix: 24
+  gateway: 192.168.1.1
+  preAllocations:
+    claim2: 192.168.0.12
+```
+
+Whenever something requires an IP address from the IPPool, it will create an
+IPClaim. The IPClaim contains a pointer to the IPPool and an owner reference
+to the object that created it.
+
+The following is an example of an IPClaim:
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPClaim
+metadata:
+  name: claim1
+spec:
+  pool:
+    Name: pool1
+status:
+  address:
+    Name: pool1-192-168-0-13
+```
+
+The controller will then reconcile this object and allocate an IP address. It
+will create an IPAddress object representing the allocated address. It will
+then update the IPPool status to list the IP Address and the IPClaim status
+to point to the IPAddress.
+
+The following is an example of an IPAddress:
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPAddress
+metadata:
+  name: pool1-192-168-0-13
+spec:
+  pool:
+    Name: pool1
+  claim:
+    Name: claim1
+  address: 192.168.0.13
+  prefix: 24
+  gateway: 192.168.0.1
+```
+
+After this allocation, the IPPool will be looking like :
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: pool1
+spec:
+  clusterName: cluster1
+  pools:
+    - start: 192.168.0.10
+      end: 192.168.0.30
+      prefix: 25
+      gateway: 192.168.0.1
+    - subnet: 192.168.1.1/26
+    - subnet: 192.168.1.128/25
+  prefix: 24
+  gateway: 192.168.1.1
+  preAllocations:
+    claim2: 192.168.0.12
+status:
+  indexes:
+    claim1: 192.168.0.13
+    claim2: 192.168.0.12
+```
+
+## Use in Metal3
+
+The IP Address Manager is used in Metal3 together with the metadata and network
+data templates feature. Each Metal3Machine (M3M) and Metal3MachineTemplate
+(M3MT) is associated with a Metal3DataTemplate that contains a metadata and /
+or a network data template that will be provided to Ironic. Those templates
```

Comment:
It's a little confusing to talk about the templates being given to Ironic. We really give the rendered data, and Ironic doesn't know about the template or any of the k8s resources.

```suggestion
or a network data template used to create the configuration for each host. Those templates
```

Reply:
fixed, thanks